### PR TITLE
Batched Gemm with output transpose

### PR DESCRIPTION
**Pull request by zjing14**
Rely on #288

### CODE REVIEW

File: library/include/ck/library/reference_tensor_operation/cpu/reference_batched_gemm_transpose.hpp [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+                float v_acc = 0;
+
+                for(int k = 0; k < K; ++k)
+                {
+                    float v_a;
+                    float v_b;
+
+                    arg.a_element_op_(v_a, static_cast<const float>(arg.a_g_m_k_(g, m, k)));
+                    arg.b_element_op_(v_b, static_cast<const float>(arg.b_g_k_n_(g, k, n)));
+
```
Comment (ltqin) [inline library/include/ck/library/reference_tensor_operation/cpu/reference_batched_gemm_transpose.hpp:70]: change static_cast to ck::type_convert

File: include/ck/tensor_operation/gpu/device/device_grouped_gemm_transpose_xdl.hpp [hunk 2]
```diff
@@ -0,0 +1,677 @@
...
+            }
+            else if constexpr(is_same<tensor_layout::gemm::ColumnMajor, ALayout>::value)
+            {
+                return make_naive_tensor_descriptor(make_tuple(M, K), make_tuple(I1, StrideA));
+            }
+        }();
+
+        if constexpr(GemmSpec == GemmSpecialization::MNPadding)
+        {
+            const auto PadM = (MPerBlock - M % MPerBlock) % MPerBlock;
```
Comment (ltqin) [inline include/ck/tensor_operation/gpu/device/device_grouped_gemm_transpose_xdl.hpp:144]: padM?

File: include/ck/tensor_operation/gpu/device/device_grouped_gemm_transpose_xdl.hpp [hunk 3]
```diff
@@ -0,0 +1,677 @@
...
+            }
+            else if constexpr(is_same<tensor_layout::gemm::ColumnMajor, BLayout>::value)
+            {
+                return make_naive_tensor_descriptor(make_tuple(K, N), make_tuple(I1, StrideB));
+            }
+        }();
+
+        if constexpr(GemmSpec == GemmSpecialization::MNPadding)
+        {
+            const auto PadN = (NPerBlock - N % NPerBlock) % NPerBlock;
```
Comment (ltqin) [inline include/ck/tensor_operation/gpu/device/device_grouped_gemm_transpose_xdl.hpp:183]: padN?

File: include/ck/tensor_operation/gpu/device/device_grouped_gemm_transpose_xdl.hpp [hunk 4]
```diff
@@ -0,0 +1,677 @@
...
+        }();
+
+        const index_t M = M0 * M1;
+        const index_t N = N0 * N1;
+
+        if constexpr(GemmSpec == GemmSpecialization::MNPadding)
+        {
+            const auto PadM = (MPerBlock - M % MPerBlock) % MPerBlock;
+            const auto PadN = (NPerBlock - N % NPerBlock) % NPerBlock;
+
```
Comment (ltqin) [inline include/ck/tensor_operation/gpu/device/device_grouped_gemm_transpose_xdl.hpp:230]: padM, padN

File: include/ck/tensor_operation/gpu/device/device_batched_gemm_transpose_xdl.hpp [hunk 5]
```diff
@@ -0,0 +1,621 @@
...
+                return make_naive_tensor_descriptor(make_tuple(M, K), make_tuple(StrideA, I1));
+            }
+            else if constexpr(is_same<tensor_layout::gemm::ColumnMajor, ALayout>::value)
+            {
+                return make_naive_tensor_descriptor(make_tuple(M, K), make_tuple(I1, StrideA));
+            }
+        }();
+
+        const auto PadM = (MPerBlock - M % MPerBlock) % MPerBlock;
+
```
Comment (ltqin) [inline include/ck/tensor_operation/gpu/device/device_batched_gemm_transpose_xdl.hpp:178]: padM

File: include/ck/tensor_operation/gpu/device/device_batched_gemm_transpose_xdl.hpp [hunk 6]
```diff
@@ -0,0 +1,621 @@
...
+            {
+                return make_naive_tensor_descriptor(make_tuple(K, N), make_tuple(StrideB, I1));
+            }
+            else if constexpr(is_same<tensor_layout::gemm::ColumnMajor, BLayout>::value)
+            {
+                return make_naive_tensor_descriptor(make_tuple(K, N), make_tuple(I1, StrideB));
+            }
+        }();
+
+        const auto PadN = (NPerBlock - N % NPerBlock) % NPerBlock;
```
Comment (ltqin) [inline include/ck/tensor_operation/gpu/device/device_batched_gemm_transpose_xdl.hpp:206]: padN

File: include/ck/tensor_operation/gpu/device/device_batched_gemm_transpose_xdl.hpp [hunk 7]
```diff
@@ -0,0 +1,621 @@
...
+                                                          make_merge_transform(make_tuple(N0, N1))),
+                                               make_tuple(Sequence<0, 1>{}, Sequence<2, 3>{}),
+                                               make_tuple(Sequence<0>{}, Sequence<1>{}));
+        }();
+
+        const index_t M = M0 * M1;
+        const index_t N = N0 * N1;
+
+        const auto PadM = (MPerBlock - M % MPerBlock) % MPerBlock;
+        const auto PadN = (NPerBlock - N % NPerBlock) % NPerBlock;
```
Comment (ltqin) [inline include/ck/tensor_operation/gpu/device/device_batched_gemm_transpose_xdl.hpp:242]: padM padN

File: include/ck/tensor_operation/gpu/device/device_batched_gemm.hpp [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+struct DeviceBatchedGemmTranspose : public BaseOperator
+{
+    virtual std::unique_ptr<BaseArgument> MakeArgumentPointer(const void* p_a,
+                                                              const void* p_b,
+                                                              void* p_c,
+                                                              GemmTransposeDesc gemm_transpose_desc,
+                                                              AElementwiseOperation a_element_op,
+                                                              BElementwiseOperation b_element_op,
+                                                              CElementwiseOperation c_element_op,
+                                                              ck::index_t BatchCount = 1) = 0;
```
Comment (asroy) [inline include/ck/tensor_operation/gpu/device/device_batched_gemm.hpp:23]: BatchCount is always needed

File: include/ck/tensor_operation/gpu/device/device_batched_gemm_transpose_xdl.hpp [hunk 9]
```diff
@@ -0,0 +1,666 @@
...
+          typename BBlockTransferSrcAccessOrder,
+          ck::index_t BBlockTransferSrcVectorDim,
+          ck::index_t BBlockTransferSrcScalarPerVector,
+          ck::index_t BBlockTransferDstScalarPerVector_K1,
+          bool BBlockLdsAddExtraN,
+          index_t CShuffleMXdlPerWavePerShuffle,
+          index_t CShuffleNXdlPerWavePerShuffle,
+          typename CDEBlockTransferClusterLengths_MBlock_MPerBlock_NBlock_NPerBlock,
+          index_t CDEBlockTransferScalarPerVector_NPerBlock,
+          ck::index_t NumPrefetch = 1,
```
Comment (asroy) [inline include/ck/tensor_operation/gpu/device/device_batched_gemm_transpose_xdl.hpp:160]: NumPrefetch should be explicitly specified.

Please make sure the ordering of these tuning parameter is the same as the non-batched version so there is less effort in refactoring which is needed very soon

File: include/ck/tensor_operation/gpu/device/device_batched_gemm_transpose_xdl.hpp [hunk 10]
```diff
@@ -0,0 +1,666 @@
...
+
+            const index_t grid_size =
+                arg.block_2_ctile_map_.CalculateGridSize(arg.c_grid_desc_m_n_) * arg.BatchCount_;
+
+            const auto K =
+                arg.a_grid_desc_k0_m_k1_.GetLength(I0) * arg.a_grid_desc_k0_m_k1_.GetLength(I2);
+
+            float ave_time = 0;
+
+            if(GridwiseGemm::CalculateHasMainKBlockLoop(K))
```
Comment (asroy) [inline include/ck/tensor_operation/gpu/device/device_batched_gemm_transpose_xdl.hpp:493]: Can be simplified using lambda

https://github.com/ROCmSoftwarePlatform/composable_kernel/blob/56adf7e9cc4fcf6592151281a727e96b625bc54f/include/ck/tensor_operation/gpu/device/device_gemm_multiple_d_xdl_cshuffle.hpp#L575-L623

File: library/include/ck/library/reference_tensor_operation/cpu/reference_batched_gemm_transpose.hpp [hunk 11]
```diff
@@ -0,0 +1,141 @@
+#ifndef REFERENCE_BATCHED_GEMM_TRANSPOSE_HPP
```
Comment (asroy) [inline library/include/ck/library/reference_tensor_operation/cpu/reference_batched_gemm_transpose.hpp:1]: I think we don't need a reference operator for every DeviceOp.

Here is an example

https://github.com/ROCmSoftwarePlatform/composable_kernel/blob/56adf7e9cc4fcf6592151281a727e96b625bc54f/example/04_gemm_add_add_fastgelu/gemm_add_add_fastgelu_xdl_fp16.cpp#L212-L237

**Review by asroy — CHANGES_REQUESTED** [review]
Please refer to comments above for suggested changes

Also, we would need instances file for batch_gemm+output_permutation soon

File: include/ck/tensor_operation/gpu/device/device_batched_gemm_transpose_xdl.hpp [hunk 12]
```diff
@@ -0,0 +1,666 @@
...
+          ck::index_t BBlockTransferSrcScalarPerVector,
+          ck::index_t BBlockTransferDstScalarPerVector_K1,
+          bool BBlockLdsAddExtraN,
+          index_t CShuffleMXdlPerWavePerShuffle,
+          index_t CShuffleNXdlPerWavePerShuffle,
+          typename CDEBlockTransferClusterLengths_MBlock_MPerBlock_NBlock_NPerBlock,
+          index_t CDEBlockTransferScalarPerVector_NPerBlock,
+          ck::index_t NumPrefetch = 1,
+          LoopScheduler LoopSched = make_default_loop_scheduler()>
+struct DeviceBatchedGemmTransposeXdl : public DeviceBatchedGemmTranspose<AElementwiseOperation,
```
Comment (asroy) [inline include/ck/tensor_operation/gpu/device/device_batched_gemm_transpose_xdl.hpp:162]: DeviceBatchedGemmCPermutation_xdl

File: library/include/ck/library/reference_tensor_operation/cpu/reference_batched_gemm_transpose.hpp (hunk 11 above)
Comment (zjing14) [reply library/include/ck/library/reference_tensor_operation/cpu/reference_batched_gemm_transpose.hpp:1]: Let us leave it to unifying Gemm kernels.

**Comment by zjing14** [comment]
@asroy  Fixed all comments, please review it again.

File: include/ck/tensor_operation/gpu/device/device_batched_gemm_c_permutation.hpp [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+#include "device_base.hpp"
+
+namespace ck {
+namespace tensor_operation {
+namespace device {
+
+template <typename AElementwiseOperation,
+          typename BElementwiseOperation,
+          typename CElementwiseOperation>
+struct DeviceBatchedGemmCPermutation : public BaseOperator
```
Comment (asroy) [inline include/ck/tensor_operation/gpu/device/device_batched_gemm_c_permutation.hpp:14]: the group GEMM version is called "Permute", here is "Permutation". Let's be consistent in naming

File: example/24_grouped_gemm_bias_transpose/grouped_gemm_bias_transpose_xdl_fp16.cpp [hunk 14]
```diff
@@ -0,0 +1,334 @@
+#include <iostream>
```
Comment (asroy) [inline example/24_grouped_gemm_bias_transpose/grouped_gemm_bias_transpose_xdl_fp16.cpp:1]: filenam and folder name should be "permute" or "permuation"

File: example/25_batched_gemm_transpose/batched_gemm_transpose_xdl_fp16.cpp [hunk 15]
```diff
@@ -0,0 +1,237 @@
+#include <iostream>
```
Comment (asroy) [inline example/25_batched_gemm_transpose/batched_gemm_transpose_xdl_fp16.cpp:1]: filenam and folder name should be "permute" or "permuation"

File: include/ck/tensor_operation/gpu/device/device_batched_gemm_c_permutation.hpp [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+{
+    virtual std::unique_ptr<BaseArgument> MakeArgumentPointer(const void* p_a,
+                                                              const void* p_b,
+                                                              void* p_c,
+                                                              GemmTransposeDesc gemm_transpose_desc,
+                                                              AElementwiseOperation a_element_op,
+                                                              BElementwiseOperation b_element_op,
+                                                              CElementwiseOperation c_element_op,
+                                                              ck::index_t BatchCount = 1) = 0;
```
Comment (asroy) [inline include/ck/tensor_operation/gpu/device/device_batched_gemm_c_permutation.hpp:23]: This is batch GEMM, so BatchCount should always be used

File: include/ck/tensor_operation/gpu/device/device_batched_gemm_c_permutation_xdl.hpp [hunk 17]
```diff
@@ -0,0 +1,638 @@
...
+          typename BElementwiseOperation,
+          typename CElementwiseOperation,
+          typename ComputePtrOffsetOfBatch,
+          typename Block2CTileMap,
+          bool HasMainKBlockLoop>
+__global__ void
+#if CK_USE_LAUNCH_BOUNDS
+    __launch_bounds__(CK_MAX_THREAD_PER_BLOCK, CK_MIN_BLOCK_PER_CU)
+#endif
+        kernel_batched_gemm_transpose_xdl(const FloatAB* __restrict__ p_a_grid,
```
Comment (asroy) [inline include/ck/tensor_operation/gpu/device/device_batched_gemm_c_permutation_xdl.hpp:60]: permuate

File: include/ck/tensor_operation/gpu/device/device_batched_gemm_c_permutation.hpp [hunk 18]
```diff
@@ -0,0 +1,38 @@
+#pragma once
```
Comment (asroy) [inline include/ck/tensor_operation/gpu/device/device_batched_gemm_c_permutation.hpp:1]: filename should use "permute"

File: library/include/ck/library/reference_tensor_operation/cpu/reference_gemm_bias_transpose.hpp [hunk 19]
```diff
@@ -0,0 +1,140 @@
+#pragma once
```
Comment (asroy) [inline library/include/ck/library/reference_tensor_operation/cpu/reference_gemm_bias_transpose.hpp:1]: filename should use "permute"

File: include/ck/tensor_operation/gpu/device/device_batched_gemm.hpp (hunk 8 above)
Comment (asroy) [reply include/ck/tensor_operation/gpu/device/device_batched_gemm.hpp]: still not fixed

File: library/include/ck/library/reference_tensor_operation/cpu/reference_batched_gemm_transpose.hpp (hunk 1 above)
Comment (asroy) [reply library/include/ck/library/reference_tensor_operation/cpu/reference_batched_gemm_transpose.hpp:70]: still not fixed

**Review by asroy — CHANGES_REQUESTED** [review]
naming changes.
static_cast issue.

**Comment by asroy** [comment]
@zjing14 can we just use this PR, instead of #288?

I've made some comments in #288 too, but you can address them in this PR

**Comment by zjing14** [comment]
Creating a new PR without grouped gemm PR dependency

**Pull request closed without merging**